scratch-debugger/debug.sh

```diff
@@ -0,0 +1,179 @@
+if [[ $# == 0 ]]; then
+  echo >&2 "USAGE: $0 POD_NAME [POD_NAMESPACE CONTAINER_NAME]"
+  exit 1
+fi
+
+# Customizable parameters
+TMP_SUBDIR="${TMP_SUBDIR:-debug-tools}"
+CONTEXT="${KUBECONTEXT:-}"
+DEBUGGER_NAME="${DEBUGGER_NAME:-debugger}"
+ARCH="${ARCH:-amd64}"
+
+# Name & namespace of target pod
+NAME=$1
+NAMESPACE=${2:-}
+CONTAINER_NAME=${3:-}
+
+# Internal variables
+KUBECTL="kubectl"
+[ -z "$CONTEXT" ] || KUBECTL="$KUBECTL --context=$CONTEXT"
+[ -z "$NAMESPACE" ] || KUBECTL="$KUBECTL --namespace=$NAMESPACE"
+
+NAMESPACE=${NAMESPACE:-default}
+
+TARGET_CNAME=${CONTAINER_NAME}
+if [ -z "$CONTAINER_NAME" ]; then
+  CONTAINER_NAME=$(${KUBECTL} get pod $NAME -o jsonpath='{.spec.containers[0].name}')
+fi
+
+INSTALL_DIR="/tmp/${TMP_SUBDIR}"
+CONTAINER_ID=$(${KUBECTL} get pod $NAME -o jsonpath="{.status.containerStatuses[?(@.name==\"$CONTAINER_NAME\")].containerID}")
+RUNTIME=${CONTAINER_ID%://*}
+CONTAINER_ID=${CONTAINER_ID#*://}
+NODE=$(${KUBECTL} get pod $NAME -o jsonpath='{.spec.nodeName}')
+
+if [[ $RUNTIME != docker ]]; then
+  echo >&2 "Error: $0 only works with a docker runtime. Found: $CONTAINER_ID"
+  exit 1
+fi
+
+# Construct the command to debug the target container.
+DEBUG_CMD="${KUBECTL} exec -i -t ${NAME}"
+[ -z "$TARGET_CNAME" ] || DEBUG_CMD="$DEBUG_CMD -c $TARGET_CNAME"
+DEBUG_CMD="$DEBUG_CMD -- ${INSTALL_DIR}/sh -c 'PATH=\$PATH:${INSTALL_DIR} sh'"
+
+if $KUBECTL exec ${NAME} -c $CONTAINER_NAME -- ${INSTALL_DIR}/echo &>/dev/null; then
+  echo "Debug tools already installed. Dumping you into the pod container now."
+  eval "$DEBUG_CMD"
+  exit 0
+fi
+
+cat <<EOF
+Debug Target Container:
+  Pod:          $NAME
+  Namespace:    $NAMESPACE
+  Node:         $NODE
+  Container:    $CONTAINER_NAME
+  Container ID: $CONTAINER_ID
+  Runtime:      $RUNTIME
+
+  "Installing busybox to $INSTALL_DIR ..."
+EOF
+
+# Cleanup the debugger pod.
+function cleanup() {
+  if ${KUBECTL} get pod ${DEBUGGER_NAME} &>/dev/null; then
+    ${KUBECTL} delete pod ${DEBUGGER_NAME}
+  fi
+}
+
+case $ARCH in
+  "amd64")
+    IMAGE=busybox
+    ;;
+  "arm")
+    IMAGE=armhf/busybox
+    ;;
+  "arm64")
+    IMAGE=aarch64/busybox
+    ;;
+  "ppc64le")
+    IMAGE=ppc64le/busybox
+    ;;
+  "s390x")
+    IMAGE=s390x/busybox
+    ;;
+  *)
+    echo "Unsupported architecture: $ARCH"
+    exit 1
+esac
+
+if  ${KUBECTL} get pod ${DEBUGGER_NAME} &>/dev/null; then
+  echo 2> "Error: pod ${DEBUGGER_NAME} is already running in namespace ${NAMESPACE}"
+  exit 1
+fi
+
+DOCKERCMD="/mnt/rootfs/usr/bin/docker -H unix:///run/docker.sock"
+
+${KUBECTL} create -f - <<EOF
+apiVersion: v1
+kind: Pod
+metadata:
+  name:        ${DEBUGGER_NAME}
+  namespace:   ${NAMESPACE}
+spec:
+  nodeName:    ${NODE}
+  restartPolicy: Never
+  containers:
+    - name:    debugger
+      image:   ${IMAGE}
+      securityContext:
+        privileged: true
+      command:
+        - sh
+        - -c
+        - "set -x; mkdir -p ${INSTALL_DIR} && \
+           ${DOCKERCMD} cp /tmp ${CONTAINER_ID}:/ && \
```

Comment:
Commented each line. This specific one is funny, since there is no mkdir in the scratch container to create the target directory structure.